### PR TITLE
Fixes a number of bugs in silences

### PR DESCRIPTION
**Pull request by grobinson-grafana**
#### What this PR does

This commit fixes the following bugs in silences:

- https://github.com/prometheus/alertmanager/pull/3877
- https://github.com/prometheus/alertmanager/pull/3898
- https://github.com/prometheus/alertmanager/pull/3897

which could cause an existing silence to be deleted/expired when updating the silence failed. This can happen when the replacing silence exceeded limits or was invalid.

#### Which issue(s) this PR fixes or relates to

Here are some screenshots showing the broken behavior, where an existing silence is expired because the limit was reached:

Screenshot 2024-06-26 at 17 05 02
Screenshot 2024-06-26 at 17 06 07
Screenshot 2024-06-26 at 17 06 18

And here is the same update performed with the fix:

Screenshot 2024-06-26 at 17 11 10
Screenshot 2024-06-26 at 17 11 18
Screenshot 2024-06-26 at 17 11 25

You can see the original silence is kept. This has also been tested for invalid silences and silences which are too large (exceed maximum size limit):

Screenshot 2024-06-26 at 17 24 12
Screenshot 2024-06-26 at 17 24 16

#### Checklist

- [x] Tests updated.
- [x] Documentation added.
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`.
- [x] [`about-versioning.md`](https://github.com/grafana/mimir/blob/main/docs/sources/mimir/configure/about-versioning.md) updated with experimental features.

### CODE REVIEW

File: go.mod
Comment (grobinson-grafana) [inline go.mod:66]: This comes from [here](https://github.com/grafana/alerting/commits/grobinson/fix-silence-fixes). It is a branch that has two commits on top of `835aff5` to avoid bringing in the changes from #8515 while it is still in review.

Comment (grobinson-grafana) [reply go.mod:66]: @santihernandezc asked me why do this instead of using `main` from `grafana/alerting`. My reasoning for this was so I don't have to bring in changes from #8515 for the bug fix as I don't know if and when that PR will be approved. If you consider this to be unnecessary I will happily update the PR to use `main`.

**Review by santihernandezc — APPROVED** [review]
LGTM 🎉

**Review by 56quarters — APPROVED** [review]
Approving since this touches `integration/` which requires a maintainer approval. I am relying on alerting team members to review the rest of the changes.

**Review by stevesg — APPROVED** [review]

**Pull request merged**
